### PR TITLE
External request harvesting optimisation

### PR DESCRIPTION
**Pull request by seran**

### CODE REVIEW

**Review by man-zhang — CHANGES_REQUESTED** [review]
Hi, Seran, thanks! might add tests (unit test or e2e) to check if the strategy is applied, ie, the response is assigned as expected based on the strategy

File: e2e-tests/spring-rest-openapi-v3/src/main/kotlin/com/foo/rest/examples/spring/openapi/v3/wiremock/harvestresponse/WmHarvestResponseRest.kt [hunk 1]
```diff
@@ -89,9 +89,9 @@ class WmHarvestResponseRest {
             if (list.size >= 2 && list.any { it is Map<*, *> && it.size > 10 })
                 return ResponseEntity.ok("Found harvested response")
             return ResponseEntity.ok("Cannot find harvested response")
-        } catch (e: Exception) {
-            return ResponseEntity.status(400).build()
-        }
+//        } catch (e: Exception) {
```
Comment (man-zhang) [inline e2e-tests/spring-rest-openapi-v3/src/main/kotlin/com/foo/rest/examples/spring/openapi/v3/wiremock/harvestresponse/WmHarvestResponseRest.kt:92]: same as the previous

**Review by man-zhang — CHANGES_REQUESTED** [review]
Hi, Seran, thanks! please check my update and comments.

File: core/src/test/kotlin/org/evomaster/core/problem/external/service/ExternalHarvesterStrategyUnitTest.kt [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        val config = EMConfig()
+        config.externalRequestResponseSelectionStrategy = EMConfig.ExternalRequestResponseSelectionStrategy.EXACT
+        config.probOfHarvestingResponsesFromActualExternalServices = 1.0
+        config.probOfMutatingResponsesBasedOnActualResponse = 2.0
+
+        val externalHarvestActualHttpWsResponseHandler = HarvestActualHttpWsResponseHandler(config, remoteController, randomness)
+        externalHarvestActualHttpWsResponseHandler.initialize()
+
+        val requests = mutableListOf<HttpExternalServiceRequest>()
+        requests.add(HttpExternalServiceRequest(UUID.randomUUID(),"GET","https://www.google.com/","https://www.google.com/",true,UUID.randomUUID().toString(),"https://www.google.com/", mapOf(),null))
```
[the remark this answers is not in the document]
Comment (seran) [reply core/src/test/kotlin/org/evomaster/core/problem/external/service/ExternalHarvesterStrategyUnitTest.kt]: Possible. I can change the test to have a WireMock service to mock the external service.

File: core/src/test/kotlin/org/evomaster/core/problem/external/service/ExternalHarvesterStrategyTest.kt [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        val successResult: HttpWsResponseParam = externalHarvestActualHttpWsResponseHandler.getACopyOfActualResponse(resultRequest, 1.0) as HttpWsResponseParam
+        val noResponseResult = externalHarvestActualHttpWsResponseHandler.getACopyOfActualResponse(noResponseRequest, 1.0)
+
+        val successStatus = successResult!!.status.values[successResult!!.status.index]
+
+        wm.shutdown()
+        DnsCacheManipulator.clearDnsCache()
+
+        assertEquals(successStatus, 200)
+        assertEquals(noResponseResult, null)
```
Comment (man-zhang) [inline core/src/test/kotlin/org/evomaster/core/problem/external/service/ExternalHarvesterStrategyTest.kt:146]: should this find the harvested request as well?

**Review by man-zhang — CHANGES_REQUESTED** [review]
Hi, Seran, thanks! I have a question about the test.

**Review by man-zhang — COMMENTED** [review]
Hi, Seran, thanks! I update the unit test a bit. If you do agree with the changes and the tests pass, then you can request the Andrea's review.

File: core/src/main/kotlin/org/evomaster/core/problem/externalservice/httpws/service/HarvestActualHttpWsResponseHandler.kt [hunk 4]
```diff
@@ -427,4 +446,64 @@ class HarvestActualHttpWsResponseHandler {
...
+                        }
+                    }
+                }
+            }
+        }
+        return out
+    }
+
+    private fun getURL(url: String): List<String> {
+        return url.split("::")
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/externalservice/httpws/service/HarvestActualHttpWsResponseHandler.kt:483]: i don't understand this function... how can a `getURL` return a list of strings and not a URL???

File: core/src/main/kotlin/org/evomaster/core/problem/externalservice/httpws/service/HarvestActualHttpWsResponseHandler.kt [hunk 5]
```diff
@@ -427,4 +446,64 @@ class HarvestActualHttpWsResponseHandler {
...
+    }
+
+    private fun getMethod(url: String): String {
+        return url.split("::")[0].lowercase()
+    }
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/externalservice/httpws/service/HarvestActualHttpWsResponseHandler.kt:488]: these functions are making major assumptions on the structure of the URL... replace with JDK API `URL` objects, which is safer and more complete

File: core/src/main/kotlin/org/evomaster/core/problem/externalservice/httpws/service/HarvestActualHttpWsResponseHandler.kt [hunk 6]
```diff
@@ -427,4 +446,64 @@ class HarvestActualHttpWsResponseHandler {
...
+
+    private fun getProtocol(url: String): String {
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/externalservice/httpws/service/HarvestActualHttpWsResponseHandler.kt:490]: see previous comment

File: core/src/main/kotlin/org/evomaster/core/problem/externalservice/httpws/service/HarvestActualHttpWsResponseHandler.kt [hunk 7]
```diff
@@ -427,4 +446,64 @@ class HarvestActualHttpWsResponseHandler {
...
+        return getURL(url)[1].split("/")[0].replace(":", "").lowercase()
+    }
+
+    private fun getDomain(url: String): String {
+        return getURL(url)[1].split("/")[2].lowercase()
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/externalservice/httpws/service/HarvestActualHttpWsResponseHandler.kt:495]: see previous comment

**Review by arcuri82 — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/org/evomaster/core/problem/externalservice/httpws/service/HarvestActualHttpWsResponseHandler.kt (hunk 4 above)
Comment (seran) [reply core/src/main/kotlin/org/evomaster/core/problem/externalservice/httpws/service/HarvestActualHttpWsResponseHandler.kt]: I changed the method with more meaning name and with comments.

File: core/src/main/kotlin/org/evomaster/core/problem/externalservice/httpws/service/HarvestActualHttpWsResponseHandler.kt [hunk 8]
```diff
@@ -479,30 +480,36 @@ class HarvestActualHttpWsResponseHandler {
...
 
-    private fun getURL(url: String): List<String> {
-        return url.split("::")
+    /**
+     * A request description contains information about the method, url, headers, and body.
+     * This extract the URL information as [URL] for ease of use.
+     *
+     * e.g.: GET::http://exists.local:12354/api/fzz::[]::{none}
+     */
+    private fun getURLFromRequestDescription(url: String): URL {
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/externalservice/httpws/service/HarvestActualHttpWsResponseHandler.kt:489]: i see... that is the confusing part: the input `url` has a misleading name, as it is not an URL, but rather a "request description". change name of all these inputs `url` string objects (also in following methods)

**Review by arcuri82 — CHANGES_REQUESTED** [review]

**Pull request merged**